filebeat/input/tcp/harvester.go

```diff
@@ -0,0 +1,158 @@
+package tcp
+
+import (
+	"bufio"
+	"bytes"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/elastic/beats/filebeat/harvester"
+)
+
+// Harvester represent a TCP server
+type Harvester struct {
+	sync.RWMutex
+	forwarder *harvester.Forwarder
+	config    config
+	server    net.Listener
+	clients   map[*Client]struct{}
+	wg        sync.WaitGroup
+	done      chan struct{}
+	splitFunc bufio.SplitFunc
+}
+
+// NewHarvester creates a new harvester that will forward events
+func NewHarvester(
+	forwarder *harvester.Forwarder,
+	cfg *common.Config,
+) (*Harvester, error) {
+
+	config := defaultConfig
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	server, err := net.Listen("tcp", config.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	sf := splitFunc([]byte(config.LineDelimiter))
+	return &Harvester{
+		config:    config,
+		forwarder: forwarder,
+		clients:   make(map[*Client]struct{}, 0),
+		done:      make(chan struct{}),
+		server:    server,
+		splitFunc: sf,
+	}, nil
+}
+
+// Run start and run a new TCP listener to receive new data
+func (h *Harvester) Run() error {
+	logp.Info("Started listening for incoming TCP connection on: %s:", h.config.Host)
+	for {
+		conn, err := h.server.Accept()
```

Comment:
I would expect that the prospector/input would be the server and then each harvester would be a client?

Reply:
Like we discussed in the input refactoring, I am not sure the wording harvester work in the context of UDP and TCP, I would be to rename `harvester` to `server` and use `client` for the actual remote connection. Would you be OK with that?

My reasoning is `harvester` is more active process to fetch new things (pull), which in the other hand a TCP / UDP is more passive (push).

Reply:
I'm not really concerned here about not using the name harvester but why `input != server`. From my point of view 1 input = 1 server, so if I start an input that should mean the server is started. Instead of having `input`, `server`, `client` I think we only need `input=server`, `client` and can remove one abstraction layer.

I'm not too concerned about not having the name harvester anywhere as this is not something exposed to the user.

Reply:
@ph Now looking at my own UDP code I see that I used also the harvester as the server :-( Leading by bad example? :-(

Reply:
Not, you lead by good example you see the problem :)